src/PSRule.Rules.Azure/rules/Azure.AKS.Rule.ps1

```diff
@@ -173,6 +173,50 @@ Rule 'Azure.AKS.CNISubnetSize' -Type 'Microsoft.ContainerService/managedClusters
     }
 } -Configure @{ AZURE_AKS_CNI_MINIMUM_CLUSTER_SUBNET_SIZE = 23 }
 
+# Synopsis: AKS clusters deployed with virtual machine scale sets should use availability zones in supported regions for high availability.
+Rule 'Azure.AKS.AvailabilityZone' -Type 'Microsoft.ContainerService/managedClusters', 'Microsoft.ContainerService/managedClusters/agentPools' -Tag @{ release = 'GA'; ruleSet = '2021_09'; } {
+    $agentPools = @(GetAgentPoolProfiles);
+
+    if ($agentPools.Length -eq 0 -or [string]::IsNullOrEmpty($TargetObject.Location)) {
+        return $Assert.Pass();
+    }
+
+    $configurationZones = $Configuration.AZURE_AKS_ADDITIONAL_REGION_AVAILABILITY_ZONE_LIST;
+
+    $virtualMachineScaleSetProvider = [PSRule.Rules.Azure.Runtime.Helper]::GetResourceType('Microsoft.Compute', 'virtualMachineScaleSets');
+
+    if ($configurationZones.Length -gt 0) {
+        $availabilityZones = @($configurationZones) + $virtualMachineScaleSetProvider.ZoneMappings;
+    }
+    else {
+        $availabilityZones = $virtualMachineScaleSetProvider.ZoneMappings;
+    }
+
+    $location = GetNormalLocation -Location $TargetObject.Location;
+
+    $locationAvailabilityZones = $availabilityZones 
+    | Where-Object { (GetNormalLocation -Location $_.Location) -eq $location } 
+    | Select-Object -ExpandProperty Zones -First 1;
+
+    if (-not $locationAvailabilityZones) {
+        return $Assert.Pass();
+    }
+
+    $joinedZoneString = $locationAvailabilityZones -join ', ';
+
+    foreach ($agentPool in $agentPools) {
+
+        # Availability zones only available on virtual machine scale sets
+        if ($Assert.HasFieldValue($agentPool, 'type', 'VirtualMachineScaleSets').Result) {
+            $validZone = $Assert.HasField($agentPool, 'availabilityZones').Result -and -not $Assert.NullOrEmpty($agentPool, 'availabilityZones').Result;
+            $Assert.Create($validZone, ($LocalizedData.AKSAvailabilityZone -f $agentPool.name, $location, $joinedZoneString));
+        }
+        else {
+            $Assert.Pass();
+        }
+    }
+} -Configure @{ AZURE_AKS_ADDITIONAL_REGION_AVAILABILITY_ZONE_LIST = @() }
```

Comment:
I probably need to test this with some custom zones. Need to check if my format in `/docs/setup/configuring-rules.md` will get serialized properly into an array of `PSCustomObject` with yamldotnet.

Reply:
Yeah, we should ideally have a test case for custom configuration. Objects should be serialized as PSObject.

Reply:
Thanks @BernieWhite. Is there a way to easily test this? Or do I have to create a `ps-rule.yml` file with these custom zones. Would be cool to be able to pass in these configurations via a `-Configuration` switch that accepts a hashtable or something.

Reply:
I know in pester I can use `TestDrive` to create files just for testing, perhaps I could use that here and dump some YAML config in there.

Reply:
@ArmaanMcleod There is a few ways to provide options:

- Pass a hashtable. See [example](https://github.com/Azure/PSRule.Rules.Azure/blob/6ecd4e7cd369f4261780a823df6826241681acc7/tests/PSRule.Rules.Azure.Tests/Azure.Convention.Tests.ps1#L49-L52)
- Pass a path to a options file. [see](https://microsoft.github.io/PSRule/concepts/PSRule/en-US/about_PSRule_Options.html)
- Pass an options object. [example](https://github.com/Azure/PSRule.Rules.Azure/blob/6ecd4e7cd369f4261780a823df6826241681acc7/tests/PSRule.Rules.Azure.Tests/Azure.Resource.Tests.ps1#L92-L100)

Reply:
Thanks @BernieWhite. Hashtable works fine. However when I use this config:

```yaml
# PSRule options for unit testing

configuration:
  AZURE_SUBSCRIPTION:
    subscriptionId: 'ffffffff-ffff-ffff-ffff-ffffffffffff'
    tenantId: 'ffffffff-ffff-ffff-ffff-ffffffffffff'
    displayName: 'Unit Test Subscription'
    state: 'NotDefined'
  AZURE_RESOURCE_GROUP:
    name: 'ps-rule-test-rg'
    location: 'eastus'
    tags:
      env: 'prod'
    properties:
      provisioningState: 'Succeeded'
  AZURE_AKS_ADDITIONAL_REGION_AVAILABILITY_ZONE_LIST:
  - location: 'Australia Southeast'
    zones:
      - "1"
      - "2"
      - "3"
  - location: 'Norway East'
    zones:
      - "1"
      - "2"
      - "3"
```

Doesn't seem to work and can't read the `$_.Location` property on line 197. 

I am just testing this with:

```pwsh
$configPath = Join-Path -Path $here -ChildPath 'ps-rule-options.yaml';
$result = Invoke-PSRule @invokeParams -InputPath $dataPath -Option $configPath
```

Seems to me its not liking the YAML format.

Reply:
@ArmaanMcleod Seems to be a bug. The object is being deserialized as a dictionary.

```powershell
$o = New-PSRuleOption .\tests\PSRule.Rules.Azure.Tests\ps-rule-options.yaml
$o.Configuration['AZURE_AKS_ADDITIONAL_REGION_AVAILABILITY_ZONE_LIST']
```

returns:

```text
Key      Value
---      -----
location Australia Southeast
zones    {1, 2, 3}
location Norway East
zones    {1, 2, 3}
```

Reply:
@BernieWhite Yeah I was able to reproduce the same just then. I guess the workaround could be to convert the dictionary to an array inside the rule, or change the YAML fomat to something else. I stuck with the array since that's how the API exports the provider zoneMappings, and just kept it consistent for the option.

Reply:
@ArmaanMcleod Try using pre-release v1.7.0-B2108021 in 15mins. Just pushing a release now.

Reply:
@BernieWhite That was quick, thanks for that 😄 . I've tested the pre-release locally and it seems to be working fine with the tests I've added. Will see if it passes CI checks.

Reply:
Yeah tests failing in CI since its using latest stable 1.6.0 PSRule module.